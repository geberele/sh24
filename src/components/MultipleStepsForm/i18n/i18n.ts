import { I18N } from './i18n.types'

export const en: I18N = {
  question1: 'What is your name?',
  question2: 'What is your email address?',
  question3: 'What service are you here for?',
  options1: 'STI Testing',
  options2: 'Contraception',
  options3: 'Other',
  labelName: 'Name',
  labelEmail: 'Email address',
  previewTitle: 'Preview',
  prev: 'Back',
  next: 'Next',
  errorMessage1: 'Please insert a valid name (2-25 characters)',
  errorMessage2: 'Please insert a valid email',
  errorMessage3: 'Please select an option',
}

export const it: I18N = {
  question1: 'Come ti chiami?',
  question2: 'Quale Ã¨ la tua email?',
  question3: 'Che tipo di servizio ti interessa?',
  options1: 'Test STI',
  options2: 'Contraccettivo',
  options3: 'Altro',
  labelName: 'Nome',
  labelEmail: 'Email',
  previewTitle: 'Preview',
  prev: 'Indietro',
  next: 'Avanti',
  errorMessage1: 'Inserisci un nome valido (2-25 lettere)',
  errorMessage2: 'Inserisci un\'email valida',
  errorMessage3: 'Seleziona un\'opzione per favore',
}
