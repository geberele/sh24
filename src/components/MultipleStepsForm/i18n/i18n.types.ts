export interface I18N {
  question1: string
  question2: string
  question3: string
  options1: string
  options2: string
  options3: string
  labelName: string
  labelEmail: string
  previewTitle: string
  prev: string
  next: string
  errorMessage1: string
  errorMessage2: string
  errorMessage3: string
}
