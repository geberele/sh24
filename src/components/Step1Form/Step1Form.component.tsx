import React from 'react'
import { useForm } from 'react-hook-form'

export interface Step1FormProps {
  label: string
  question: string
  next: string
  value: string
  errorMessage: string
  nextStep: () => void
  onChange: (value: string) => void
}

export const Step1Form = ({
  label,
  question,
  next,
  value,
  nextStep,
  onChange,
  errorMessage,
}: Step1FormProps) => {
  const { register, handleSubmit, errors } = useForm()

  return (
    <form onSubmit={handleSubmit(nextStep)}>
      <h1>{question}</h1>
      <label htmlFor="name">{label}</label>
      <input
        name="name"
        ref={register({ required: true, minLength: 2, maxLength: 25 })}
        defaultValue={value}
        onChange={(e) => onChange(e.target.value)}
        data-testid="input-name"
      />

      {errors.name &&
      <div
        className="error"
        data-testid="input-error"
      >
        {errorMessage}
      </div>
      }

      <div className="navigation">
        <button
          type="submit"
          className="navigationRight"
          data-testid="button-next"
        >
          {next}
        </button>
      </div>
    </form>
  )
}
