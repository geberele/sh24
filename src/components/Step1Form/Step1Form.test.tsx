import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { fireEvent, waitFor, screen } from '@testing-library/dom'
import { Step1Form, Step1FormProps } from './Step1Form.component'
import { en } from '../MultipleStepsForm/i18n/i18n'

describe('Step1Form', () => {
  const nextStep = jest.fn()

  const props: Step1FormProps = {
    question: en.question1,
    label: en.labelName,
    next: en.next,
    value: name,
    nextStep: nextStep,
    onChange: () => {
      return
    },
    errorMessage: en.errorMessage1,
  }

  const setup = () => {
    const utils = render(<Step1Form {...props}/>)
    const input = screen.getByTestId('input-name') as HTMLInputElement
    const submit = screen.getByTestId('button-next') as HTMLInputElement
    return {
      input,
      submit,
      ...utils,
    }
  }

  it('renders', () => {
    const { asFragment } = setup()
    expect(asFragment()).toMatchSnapshot()
  })

  it('returns error if characters less than 2', async () => {
    const { input, submit } = setup()
    fireEvent.change(input, { target: { value: '2' } })
    fireEvent.click(submit)
    const error = await waitFor(() => screen.getByTestId('input-error'))
    expect(error.textContent).toBe(en.errorMessage1)
  })

  it('returns error if characters more than 25', async () => {
    const { input, submit } = setup()
    fireEvent.change(input, { target: { value: '12345678912345678912345678' } })
    fireEvent.click(submit)
    const error = await waitFor(() => screen.getByTestId('input-error'))
    expect(error.textContent).toBe(en.errorMessage1)
  })

  it('submits the form if name is valid', async () => {
    const { input, submit } = setup()
    fireEvent.change(input, { target: { value: 'Mike' } })
    fireEvent.click(submit)
    await waitFor(() => expect(nextStep).toHaveBeenCalledTimes(1))
  })
})
