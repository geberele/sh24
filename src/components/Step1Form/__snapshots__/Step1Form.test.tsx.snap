// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step1Form renders 1`] = `
<DocumentFragment>
  <form>
    <h1>
      What is your name?
    </h1>
    <label
      for="name"
    >
      Name
    </label>
    <input
      data-testid="input-name"
      name="name"
      value=""
    />
    <div
      class="navigation"
    >
      <button
        class="navigationRight"
        data-testid="button-next"
        type="submit"
      >
        Next
      </button>
    </div>
  </form>
</DocumentFragment>
`;
